Source/Core/RequestQueue.js

```diff
@@ -0,0 +1,186 @@
+define([
+        './Check',
+        './defineProperties'
+    ], function(
+        Check,
+        defineProperties) {
+    'use strict';
+
+    /**
+     * Priority queue for the {@link RequestScheduler} implemented as a sorted array.
+     * The request with the highest priority is placed at index 0 and the request
+     * with lowest priority is placed at index <code>length - 1</code>.
+     * <p>
+     * A lower <code>request.priority</code> value indicates that the request has higher priority. See {@link Request#priority}.
+     * </p>
+     *
+     * @alias RequestQueue
+     * @constructor
+     * @private
+     *
+     * @param {Number} maximumLength The maximum length of the queue.
+     */
+    function RequestQueue(maximumLength) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.number('maximumLength', maximumLength);
+        //>>includeEnd('debug');
+
+        this._array = new Array(maximumLength);
+        this._length = 0;
+        this._maximumLength = maximumLength;
+    }
+
+    defineProperties(RequestQueue.prototype, {
+        /**
+         * Gets the length of the queue.
+         *
+         * @memberof RequestQueue.prototype
+         *
+         * @type {Number}
+         * @readonly
+         */
+        length : {
+            get : function() {
+                return this._length;
+            }
+        }
+    });
+
+    /**
+     * Get the request at the given index.
+     *
+     * @param {Number} index The index of the request.
+     *
+     * @return {Request} The request at the given index.
+     */
+    RequestQueue.prototype.get = function(index) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.number.greaterThanOrEquals('index', index, 0);
+        Check.typeOf.number.lessThan('index', index, this._length);
+        //>>includeEnd('debug');
+        return this._array[index];
+    };
+
+    /**
+     * Insert a request into the queue. If the length would grow greater than the maximum length
+     * of the queue, the lowest priority request is removed and returned.
+     *
+     * @param {Request} request The request to insert.
+     *
+     * @return {Request|undefined} The request that was removed from the queue if the queue is at full capacity.
+     */
+    RequestQueue.prototype.insert = function(request) {
```

Comment:
Considering how and when this is called during a frame - you know better than me - does it make more sense to queue up inserts, sort them, then merge once per frame so there is potentially less overhand that constantly keeping this sorted?

Or maybe there is another variation - just take advantage of this being specific to requests, not a general min-max queue.

Reply:
I think the rationale for not queuing up all the requests is we can know right away if a request will not be able to be issued, and return `undefined` in `RequestScheduler.request`. This leads to a faster path in terrain/imagery/3d-tiles than if the request is issued but later cancelled. I'll have to think about tradeoffs between the two approaches.

Reply:
Ah, that can be roadmapped as part of the optimizations.